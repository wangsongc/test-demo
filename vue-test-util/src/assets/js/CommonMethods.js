/**
 * 获取当前视口的宽度，判断默认样式
 * @returns true<1025<=false
 */
export function checkWidth (size) {
  let width = size || window.document.body.offsetWidth
  if (width < 1025) {
    return true
  } else {
    return false
  }
}
