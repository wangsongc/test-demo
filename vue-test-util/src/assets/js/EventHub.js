// (C) 2018 SEIKO SOLUTIONS INC.

import Vue from 'vue'

const eventHub = new Vue()

export default eventHub
