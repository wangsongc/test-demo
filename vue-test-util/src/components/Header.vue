<template>
  <div class="header">
    <div class="collapse-btn" @click="collapseChage">
      <i class="el-icon-menu"></i>
    </div>
    <div class="logo">Unit-test-with-jest</div>
    <div class="tips">本页面用于展示一些单元测试用例，供vue项目单元测试使用，布局样式等如有BUG，请忽略</div>
    <div class="user-info">
      <span class="el-dropdown-link">
        <img class="user-logo" src="../assets/logo.png">
        {{name}}
      </span>
    </div>
  </div>
</template>
<script>
import eventHub from '@/assets/js/EventHub.js'
import { checkWidth } from '@/assets/js/CommonMethods.js'
export default {
  data () {
    return {
      collapse: checkWidth(),
      name: 'admin'
    }
  },
  computed: {},
  methods: {
    collapseChage () {
      this.collapse = !this.collapse
      eventHub.$emit('collapse', this.collapse)
    }
  }
}
</script>
<style scoped>
.header {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 70px;
  font-size: 22px;
  line-height: 70px;
  color: #fff;
}
.collapse-btn {
  float: left;
  padding: 0 21px;
  cursor: pointer;
}
.collapse-btn img {
  vertical-align: middle;
}
.collapse-btn:hover {
  background: rgb(40, 52, 70);
}
.header .logo {
  float: left;
  width: 200px;
  /* text-align: center; */
}
.header .tips {
  float: left;
}
.user-info {
  float: right;
  padding-right: 50px;
  font-size: 16px;
  color: #fff;
}
.user-info .el-dropdown-link {
  position: relative;
  display: inline-block;
  padding-left: 50px;
  color: #fff;
  cursor: pointer;
  vertical-align: middle;
}
.user-info .user-logo {
  position: absolute;
  left: 0;
  top: 15px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
}
</style>
