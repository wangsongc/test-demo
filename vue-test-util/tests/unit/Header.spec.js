/**
 * Header.vue组件
 * 单元测试包括以下
 * data v-bind
 * func
 * eventHub（$emit）
 */

import { shallowMount } from '@vue/test-utils'
import Header from '@/components/Header.vue'
import eventHub from '@/assets/js/EventHub.js'
import { checkWidth } from '@/assets/js/CommonMethods.js'

describe('Header.vue', () => {
  let wrapper
  // 每个it开始之前执行
  beforeEach(() => {
    wrapper = shallowMount(Header)
  })
  // 每个it结束时执行
  afterEach(() => {
    wrapper.destroy()
  })

  // 测试内容：data v-bind
  // 断言data中变量的默认值
  it('data test', () => {
    expect(wrapper.vm.name).toBe('admin')
    expect(wrapper.vm.collapse).toBeTruthy() // 函数checkWidth()在测试环境下的值一致都是true
    wrapper.vm.name = 'holy'
    expect(wrapper.find('.el-dropdown-link').text()).toBe('holy')
  })

  // 测试内容：data中的函数checkWidth()
  it('checkWidth test', () => {
    // window.document.body.offsetWidth的值再测试环境下一直都是0，所以checkWidth的值一致都是true，无法改变
    expect(checkWidth()).toBeTruthy()
    expect(wrapper.vm.collapse).toBeTruthy()
  })

  // 测试内容：collapseChage()
  // 点击折叠侧边栏按钮，正确触发collapseChage函数
  it('click icon func is called', () => {
    const mockFn = jest.fn()
    wrapper.setMethods({
      collapseChage: mockFn
    })
    wrapper.find('.collapse-btn').trigger('click')
    // 断言函数被触发，且被触发一次
    expect(mockFn).toBeCalled()
    expect(mockFn).toHaveBeenCalledTimes(1)
  })

  // 测试内容：collapseChage（）函数执行结果测试-eventHub
  // 函数被触发后，collapse取反且eventHub的$emit触发
  it('when collapseChage() is called', () => {
    const mockFn = jest.fn()
    eventHub.$on('collapse', mockFn)
    // 第一次触发函数
    wrapper.vm.collapseChage()
    expect(wrapper.vm.collapse).toBeFalsy()
    expect(mockFn).toBeCalled()
    expect(mockFn).toHaveBeenCalledTimes(1)
    expect(mockFn).toHaveBeenCalledWith(false)
    // 第二次出发函数
    wrapper.vm.collapseChage()
    expect(wrapper.vm.collapse).toBeTruthy()
    expect(mockFn).toBeCalled()
    expect(mockFn).toHaveBeenCalledTimes(2)
    expect(mockFn).toHaveBeenCalledWith(true)
  })

  // 测试内容：snapshot->概括的测试DOM结构
  // 如果组件内存在比较特殊的需要测试的DOM结构的话，可以单独测试（详见AppButton测试文件末尾）
  it('snapshot test', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
