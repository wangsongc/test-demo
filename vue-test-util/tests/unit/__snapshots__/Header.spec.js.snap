// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue snapshot test 1`] = `
<div class="header">
  <div class="collapse-btn"><i class="el-icon-menu"></i></div>
  <div class="logo">Unit-test-with-jest</div>
  <div class="tips">本页面用于展示一些单元测试用例，供vue项目单元测试使用，布局样式等如有BUG，请忽略</div>
  <div class="user-info"><span class="el-dropdown-link"><img src="../assets/logo.png" class="user-logo">
      admin
    </span></div>
</div>
`;
